File: spec/game/match_hud_spec.lua
local hud = require("game.match_hud")
local combat_feedback = require("game.presentation.combat_feedback")
local combat_presentation = require("game.presentation.combat")
local match_sim = require("sim.match")
local t = require("spec.support.runner")
local teams = require("data.teams")
local crowded_fixture = require("spec.fixtures.crowded_combat_feedback")

---@return MatchHudContext
local function context()
    return {
        home_name = "Nebula FC",
        away_name = "Orion Miners",
        arena_name = "Helios Crown",
        arena_location = "Kairon-9 Orbit",
        tactic_name = "Press High",
        formation_name = "1-2-1",
    }
end

---@return MatchState
local function state()
    return match_sim.new({
        home = teams.nebula,
        away = teams.orion,
        field = { w = 960, h = 540 },
    })
end

t.describe("match HUD model", function()
    t.it("formats time, identity, plan, possession, and clamped stamina", function()
        local value = state()
        value.time_left = 5.9
        value.players[value.controlled].sprint_meter = 1.4
        local model = hud.model(value, context())
        t.eq(model.clock, "0:05")
        t.eq(model.possession, "NEBULA FC POSSESSION")
        t.eq(model.possession_marker, "filled")
        t.is_true(model.player_name ~= "")
        t.is_true(model.player_detail:match("TERRAN") ~= nil)
        t.eq(model.plan, "PLAN · PRESS HIGH · 1-2-1")
        t.eq(model.stamina, 1)

        value.owner = nil
        model = hud.model(value, context())
        t.eq(model.possession, "LOOSE BALL")
        t.eq(model.possession_marker, "outline")

        value.owner = 6
        model = hud.model(value, context())
        t.eq(model.possession, "ORION MINERS POSSESSION")
        t.eq(model.player_state, "DEFENDING")
    end)

    t.it("authors kickoff, goal, replay, and full-time broadcast phases", function()
        local value = state()
        for _, phase in ipairs({ "kickoff", "goal", "replay", "full_time" }) do
            local ctx = context()
            ctx.phase = phase
            ctx.scoring_team = "home"
            local model = hud.model(value, ctx)
            t.is_true(model.announcement_title ~= nil, phase .. " needs a title")
            t.is_true(model.announcement_detail ~= nil, phase .. " needs supporting detail")
        end
    end)

    t.it("keeps every HUD region inside supported viewports", function()
        for _, size in ipairs({ { 960, 540 }, { 1280, 720 }, { 1920, 1080 }, { 1280, 800 } }) do
            local layout = hud.layout({ w = size[1], h = size[2] })
            for id, rect in pairs(layout) do
                if type(rect) == "table" then
                    t.is_true(rect.x >= 0 and rect.y >= 0, id .. " begins outside the viewport")
                    t.is_true(rect.x + rect.w <= size[1], id .. " exceeds viewport width")
                    t.is_true(rect.y + rect.h <= size[2], id .. " exceeds viewport height")
                end
            end
        end
    end)

    t.it("shows geometry-labeled equipment readiness and confirmed result feedback", function()
        local value = crowded_fixture.state
        value.controlled = 2
        local combat_model = combat_presentation.model(value, crowded_fixture.combat)
        local feedback_state = combat_feedback.new()
        local confirmed = crowded_fixture.events[1]
        combat_feedback.confirm(
            feedback_state,
            combat_feedback.accepted(confirmed.payload, confirmed.id)
        )
        local ctx = context()
        ctx.combat_enabled = true
        ctx.combat = combat_model.players[value.controlled]
        ctx.combat_notice = combat_feedback.notice(feedback_state, value.controlled)
        local model = hud.model(value, ctx)
        t.is_true(model.equipment_label ~= nil)
        t.is_true(model.equipment_state ~= nil)
        t.eq(model.feedback_text, "EQUIPMENT COMMITTED")
        t.eq(model.feedback_glyph, "chevron")
        t.is_true(model.equipment_progress >= 0 and model.equipment_progress <= 1)
    end)
end)

File: ts/packages/app/src/match_hud.spec.ts
// Ported from spec/game/match_hud_spec.lua.
//
// The Lua spec drives `hud.model` from `render_frame.hud(match_sim.new(...))`
// -- a real simulation state run through the Rust-owned RenderFrame producer
// (`render/frame.lua` -> `crates/gc-render`; v2/README.md §2) -- and from
// `combat_presentation.model`/`combat_feedback` (`@gc/presentation`, not a
// declared dependency of this package). Neither is available to this port.
//
// `hud.model` itself only transforms an already-built `RenderFrameHud` +
// `MatchHudContext` into a `MatchHudModel`; none of that transformation is
// simulation math. Every case below is therefore ported against hand-built
// `RenderFrameHud`/`CombatPlayerPresentation`/`CombatFeedbackNotice`
// fixtures standing in for what the real Rust/`@gc/presentation` pipeline
// would have produced -- same technique `@gc/screens`'s `content.ts`-based
// fixtures use, and the "supplementary unit tests against the ported
// module's own control flow" the task brief calls for. The specific team
// names/species/notice text below are chosen to exercise `hud.model`'s
// branches, not transcribed from the unreachable Lua fixture values.

import { describe, expect, it } from "vitest";
import { hud, type IdentityPort, type MatchHudContext, type RenderFrameHud } from "./match_hud.ts";

const IDENTITY: IdentityPort = {
  forPlayer: (playerId) =>
    playerId === "missing"
      ? undefined
      : { name: "Zyro Vex", species_name: "Terran", position: "forward" },
};

function context(overrides: Partial<MatchHudContext> = {}): MatchHudContext {
  return {
    home_name: "Nebula FC",
    away_name: "Orion Miners",
    arena_name: "Helios Crown",
    arena_location: "Kairon-9 Orbit",
    tactic_name: "Press High",
    formation_name: "1-2-1",
    ...overrides,
  };
}

function scoreboard(overrides: Partial<RenderFrameHud> = {}): RenderFrameHud {
  return {
    home_score: 0,
    away_score: 0,
    time_left: 5.9,
    controlled_id: "zyro_vex",
    controlled_team: "home",
    controlled_is_keeper: false,
    controlled_owns_ball: true,
    controlled_stamina: 1.4,
    species_shape: "round",
    species_color: [0.5, 0.5, 0.5],
    ...overrides,
  };
}

describe("match HUD model", () => {
  it("formats time, identity, plan, possession, and clamped stamina", () => {
    const filled = hud.model(IDENTITY, scoreboard({ possession_team: "home" }), context());
    expect(filled.clock).toBe("0:05");
    expect(filled.possession).toBe("NEBULA FC POSSESSION");
    expect(filled.possession_marker).toBe("filled");
    expect(filled.player_name).not.toBe("");
    expect(filled.player_detail).toMatch(/TERRAN/);
    expect(filled.plan).toBe("PLAN · PRESS HIGH · 1-2-1");
    expect(filled.stamina).toBe(1);

    const loose = hud.model(IDENTITY, scoreboard(), context());
    expect(loose.possession).toBe("LOOSE BALL");
    expect(loose.possession_marker).toBe("outline");

    const contested = hud.model(
      IDENTITY,
      scoreboard({ possession_team: "away", controlled_owns_ball: false }),
      context(),
    );
    expect(contested.possession).toBe("ORION MINERS POSSESSION");
    expect(contested.player_state).toBe("DEFENDING");
  });

  it("authors kickoff, goal, replay, and full-time broadcast phases", () => {
    const phases = ["kickoff", "goal", "replay", "full_time"] as const;
    for (const phase of phases) {
      const model = hud.model(
        IDENTITY,
        scoreboard(),
        context({ phase, scoring_team: "home" }),
      );
      expect(model.announcement_title, `${phase} needs a title`).toBeDefined();
      expect(model.announcement_detail, `${phase} needs supporting detail`).toBeDefined();
    }
  });

  it("keeps every HUD region inside supported viewports", () => {
    const sizes: readonly [number, number][] = [
      [960, 540],
      [1280, 720],
      [1920, 1080],
      [1280, 800],
    ];
    for (const [w, h] of sizes) {
      const layout = hud.layout({ w, h });
      for (const [id, rect] of Object.entries(layout)) {
        if (typeof rect === "object") {
          expect(rect.x >= 0 && rect.y >= 0, `${id} begins outside the viewport`).toBe(true);
          expect(rect.x + rect.w <= w, `${id} exceeds viewport width`).toBe(true);
          expect(rect.y + rect.h <= h, `${id} exceeds viewport height`).toBe(true);
        }
      }
    }
  });

  it("shows geometry-labeled equipment readiness and confirmed result feedback", () => {
    const model = hud.model(
      IDENTITY,
      scoreboard(),
      context({
        combat_enabled: true,
        combat: { family_name: "spike_launcher", readiness: "committed", cooldown_fraction: 0, phase: "windup", phase_progress: 0.4 },
        combat_notice: { text: "EQUIPMENT COMMITTED", glyph: "chevron" },
      }),
    );
    expect(model.equipment_label).toBeDefined();
    expect(model.equipment_state).toBeDefined();
    expect(model.feedback_text).toBe("EQUIPMENT COMMITTED");
    expect(model.feedback_glyph).toBe("chevron");
    expect(model.equipment_progress >= 0 && model.equipment_progress <= 1).toBe(true);
  });
});
